content/blog/throttling-coredns.md

```diff
@@ -0,0 +1,153 @@
+To provide an high incoming query burst, I set up the test to spin up 150 DNS client replicas
+which in total resulted in a total incoming request of (200 x 150 = ) 30k qps to CoreDNS.
+
+The Configuration I used to test is the default as of Kubernetes v1.14 with the exception that I used the [`erratic` plugin](https://coredns.io/plugins/erratic/)
+instead of the `forward` plugin to overcome any Network shortcomings.
+
+The CoreDNS Configuration I used is as follows: 
+```corefile
+ .:53 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+          pods insecure
+          upstream
+          fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        throttle 100
+        erratic {
+          drop 0
+        }
+        cache 30
+        loop
+        reload 2s
+        loadbalance
+    }
+```
+
+### Memory consumption of CoreDNS without the `throttle` plugin
+
+Without the `throttle` plugin, CoreDNS was crashing repeatedly due to it being `OOMKilled`. The maximum consumed memory of CoreDNS 
+that was captured before the crash is 1.655 GiB.
+
+~~~logs
+      State:          Running
+      Started:      Tue, 26 Mar 2019 12:20:48 -0400
+      Last State:     Terminated
+      Reason:       OOMKilled
+      Exit Code:    137
+      Started:      Tue, 26 Mar 2019 12:20:07 -0400
+      Finished:     Tue, 26 Mar 2019 12:20:26 -0400
+      Ready:          True
+      Restart Count:  2
+~~~ 
+
+It is worth noting that the amount of memory consumed by CoreDNS is proportional to the number of `goroutines`. 
+The metrics captured for the memory consumption and the number of `goroutines`: 
+
+![no-throttle-memory](https://user-images.githubusercontent.com/30265084/55086273-7923c700-507e-11e9-8735-ff18d407c40c.png)
+
+
+### Memory consumption of CoreDNS using the `throttle` plugin
+
+I tested the same scenario with the `throttle` plugin enabled. 
+Example usage of `throttle` in the CoreDNS Corefile configuration:
+
+~~~ corefile
+. {
+    throttle <max-inflight-queries>
+}
+~~~
+
+I had tested for three different values for the `max-inflight-queries`, where it was set to 100, 500 and 1000
+to see the memory consumption in each of the cases.
+
+In this test, the base memory used by CoreDNS when there are no incoming queries is 20MiB. 
+The table below shows the throttled value of the `max-inflight-queries` to the corresponding difference of memory from the
+base memory CoreDNS was consuming with no incoming queries and memory when there were queries from the 150 DNS client replicas.
+
+| Throttle             | Memory Usage (Δ) |  
+|:--------------------:|:----------------:|
+|  100                 |   12MiB          |
+|  500                 |   20MiB          |
+| 1000                 |   25MiB          |
+
+As seen above, the memory used by CoreDNS using the `throttle` plugin is way less and hence it does not result in OOMkills.
```

Comment:
I think you should mention the fact that throttle does drop incoming packets when the limit is exceeded.  Probably should list the number of dropped packets for each test.

Reply:
maybe not list them for each test, but at least explain that this happens.

Reply:
Yes, I should make that clear.
Although, CoreDNS without the throttle plugin, if it does not get OOMkilled will drop packets since the incoming requests from the client exceed the capacity of CoreDNS.